lib/saml_idp/xml_security.rb

```diff
@@ -77,6 +74,29 @@ def validate(idp_cert_fingerprint, soft = true, options = {})
         validate_doc(base64_cert, soft, options)
       end
 
+      def validate_with_sha256(idp_certificate, options = {})
+        if request_cert && Base64.decode64(request_cert) != idp_certificate.to_der
+          raise ValidationError.new('Request certificate not valid or registered', :request_cert_not_registered)
+        end
+
+        validate_doc(request_cert, false, options)
+      end
+
+      def request_cert
+        @request_cert ||= if cert_element.text.blank?
+          raise ValidationError.new(
+            'Certificate element present in response (ds:X509Certificate) but evaluating to nil',
+            :no_certificate_in_request
+          )
+        end
+
+        cert_element.text
```

Comment:
As-is, the memoization seems to be memoizing the `raise` which is not actually a value? so I think this is memoizing `nil`

Unless `cert_element.text` is very expensive, probably easier not to memoize at all

```suggestion
        if cert_element.text.blank?
          raise ValidationError.new(
            'Certificate element present in response (ds:X509Certificate) but evaluating to nil',
            :no_certificate_in_request
          )
        end

        cert_element.text
```

Reply:
updated in https://github.com/18F/saml_idp/pull/126/commits/1eb5cca1e7498f700556ecce5c72989fad66a218